### PR TITLE
Add Model for Inspection, ArtifactRule

### PR DESCRIPTION
**Pull request by Xynnn007**
* models for `Inspection`, `ArtifactRule`
* tests of serializing and deserializing for `Inspection`, `ArtifactRule`, `Step` and `Layout`

### CODE REVIEW

File: src/models/layout/step.rs [hunk 1]
```diff
@@ -159,9 +120,9 @@ impl Step {
     }
 
     /// Add an expected product artifact rule to this Step
-    pub fn add_expected_products(mut self, expected_product: ArtifactRule) -> Self {
+    pub fn add_expected_product(mut self, expected_product: ArtifactRule) -> Self {
         self.supply_chain_item
-            .add_expected_products(expected_product);
+            .add_expected_product(expected_product);
```
Comment (SantiagoTorres) [inline src/models/layout/step.rs:125]: Hah, I like this rename

File: src/models/layout/rule.rs [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+        let mut seq = serializer.serialize_seq(len)?;
+        for e in statement {
+            seq.serialize_element(&e)?;
+        }
+        seq.end()
+    }
+}
+
+/// Visitor helps to deserialize `ArtifactRule`
+struct ArtifactRuleVisitor {}
```
Comment (SantiagoTorres) [inline src/models/layout/rule.rs:228]: I like the use of a visitor here

File: src/models/layout/rule.rs [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+        let rule = ArtifactRuleBuilder::new()
+            .set_type("DELETE")
+            .pattern("foo.pyc")
+            .build()
+            .unwrap();
+
+        let rule_parsed: ArtifactRule = serde_json::from_str(json).unwrap();
+        assert_eq!(rule, rule_parsed);
+    }
+}
```
Comment (SantiagoTorres) [inline src/models/layout/rule.rs:629]: Broadly speaking, I like this "factory" for match rules. I think we should also at least ticketize a docs.rs pass for this or so.

**Review by SantiagoTorres — APPROVED** [review]
Could we ticketize the docs for the artifact rule module?

File: src/models/layout/rule.rs (hunk 3 above)
Comment (Xynnn007) [reply src/models/layout/rule.rs:745]: Good suggestion!  I've add some docs for it, but unfortunately rust clippy test did not pass. I didn't change anything about the error file `crypto.rs`, very weird.

Comment (SantiagoTorres) [reply src/models/layout/rule.rs:745]: Yeah, I've seen this happen before. We may refactor `crypto.rs` during gsoc so I'm not exactly worried

**Pull request merged**